File: backend/app/schemas/provider.py
from uuid import UUID
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


# --- Category ---

class CategoryOut(BaseModel):
    id: UUID
    name: str
    slug: str
    description: str | None = None
    icon: str | None = None

    model_config = {"from_attributes": True}


# --- Tag ---

class TagOut(BaseModel):
    id: UUID
    name: str
    slug: str

    model_config = {"from_attributes": True}


# --- Provider Categories / Tags (with metadata) ---

class ProviderCategoryOut(BaseModel):
    category: CategoryOut
    source: str
    confidence: float | None = None

    model_config = {"from_attributes": True}


class ProviderTagOut(BaseModel):
    tag: TagOut
    source: str

    model_config = {"from_attributes": True}


# --- Provider ---

class ProviderCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    contact_email: str | None = None
    contact_phone: str | None = None
    website: str | None = None
    logo_url: str | None = None
    city: str | None = None
    country: str | None = None
    category_ids: list[UUID] = Field(default_factory=list)
    tag_names: list[str] = Field(default_factory=list)


class ProviderUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, min_length=1)
    contact_email: str | None = None
    contact_phone: str | None = None
    website: str | None = None
    logo_url: str | None = None
    city: str | None = None
    country: str | None = None
    category_ids: list[UUID] | None = None
    tag_names: list[str] | None = None


class ProviderOut(BaseModel):
    id: UUID
    name: str
    description: str
    contact_email: str | None = None
    contact_phone: str | None = None
    website: str | None = None
    logo_url: str | None = None
    city: str | None = None
    country: str | None = None
    created_at: datetime
    updated_at: datetime
    categories: list[ProviderCategoryOut] = []
    tags: list[ProviderTagOut] = []

    model_config = {"from_attributes": True}


class ProviderListOut(BaseModel):
    id: UUID
    name: str
    description: str
    city: str | None = None
    country: str | None = None
    created_at: datetime
    categories: list[ProviderCategoryOut] = []
    tags: list[ProviderTagOut] = []

    model_config = {"from_attributes": True}


# --- Pagination ---

class PaginatedResponse(BaseModel):
    items: list[ProviderListOut]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/provider.ts
export interface Category {
  id: string
  name: string
  slug: string
  description?: string
  icon?: string
}

export interface Tag {
  id: string
  name: string
  slug: string
}

export interface ProviderCategory {
  category: Category
  source: 'manual' | 'ai'
  confidence?: number
}

export interface ProviderTag {
  tag: Tag
  source: 'manual' | 'ai'
}

export interface Provider {
  id: string
  name: string
  description: string
  contact_email?: string
  contact_phone?: string
  website?: string
  logo_url?: string
  city?: string
  country?: string
  created_at: string
  updated_at: string
  categories: ProviderCategory[]
  tags: ProviderTag[]
}

export interface ProviderCreate {
  name: string
  description: string
  contact_email?: string
  contact_phone?: string
  website?: string
  logo_url?: string
  city?: string
  country?: string
  category_ids: string[]
  tag_names: string[]
}

export type ProviderUpdate = Partial<ProviderCreate>

export interface ProviderDraftForm {
  name: string
  description: string
  contact_email: string
  contact_phone: string
  website: string
  city: string
  country: string
  category_ids: string[]
  tag_names: string[]
}

export interface PaginatedResponse {
  items: Provider[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface AppliedFilters {
  country?: string
  city?: string
  categories: string[]
  technologies: string[]
  specialties: string[]
}

export interface DirectorySearchResult {
  id: string
  name: string
  description: string
  city?: string
  country?: string
  score: number
  semantic_score: number
  lexical_score: number
  metadata_score: number
  match_reasons: string[]
}

export interface WebSearchResult {
  title: string
  snippet: string
  url: string
  source: string
  city?: string
  country?: string
  detected_categories: string[]
  detected_tags: string[]
  match_reasons: string[]
}

export interface SearchRecommendation {
  label: string
  target_type: string
  target_id: string
  reason: string
}

export interface SearchAnalysis {
  summary: string
  recommendations: SearchRecommendation[]
}

export interface SearchMeta {
  used_llm_parser: boolean
  used_web_search: boolean
  strategy: string
  web_provider?: string | null
  warning?: string | null
}

export interface SearchResponse {
  query: string
  interpreted_query: string
  applied_filters: AppliedFilters
  directory_results: DirectorySearchResult[]
  web_results: WebSearchResult[]
  analysis?: SearchAnalysis | null
  session_id: string
  meta: SearchMeta
}

export interface SearchFollowupResponse {
  answer: string
  referenced_results: string[]
  warning?: string | null
}

export interface WebProviderImportPreviewRequest {
  url: string
  regenerate?: boolean
}

export interface WebProviderImportDuplicate {
  id: string
  name: string
  website?: string | null
}

export interface WebProviderImportPreviewResponse {
  status: 'ready' | 'duplicate' | 'error'
  draft: ProviderDraftForm | null
  source_url: string
  warnings: string[]
  duplicate_provider?: WebProviderImportDuplicate | null
  provenance: string
}

export interface EnrichResult {
  provider_id: string
  categorization: {
    categories: { name: string; confidence: number }[]
    tags: string[]
  }
  entities: {
    services: string[]
    technologies: string[]
    specialties: string[]
    suggested_city?: string
    suggested_country?: string
  }
  embedding_generated: boolean
}
